webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeConfigurationService.scala

```diff
@@ -0,0 +1,350 @@
+    }
+    )
+  )
+  // start updating
+  updateCacheFromRequest.forever.forkDaemon.runNow
+  /**
+   * Clear cache. Try a reload asynchronously, disregarding
+   * the result
+   */
+  override def clearCache(): Unit = {
+    init().runNow
+    logEffect.debug("Node expected reports cache cleared")
+  }
+
+
+  /**
+   * Do something with the action we received
+   */
+  private[this] def performAction(action: CacheComplianceQueueAction): IOResult[Unit] = {
+    import CacheComplianceQueueAction._
+    // in a semaphore
+    semaphore.withPermit(
+       action match {
+                          case insert: InsertNodeInCache => IOResult.effectNonBlocking {  cache = cache + (insert.nodeId -> None) }
+                          case delete: RemoveNodeInCache => IOResult.effectNonBlocking {  cache = cache.removed(delete.nodeId) }
+                          case update: UpdateNodeConfiguration => IOResult.effectNonBlocking {  cache = cache + (update.nodeId -> Some(update.nodeConfiguration)) }
+                          case something =>
+                            Inconsistency(s"NodeConfiguration service cache received unknown command : ${something}").fail
+                           // should not happen
+                   }
+    )
+  }
+
+  /**
+   * invalidate with an action to do something
+   * order is important
+   */
+  def invalidateWithAction(actions: Seq[(NodeId, CacheComplianceQueueAction)]): IOResult[Unit] = {
+    ZIO.when(actions.nonEmpty) {
+      ReportLoggerPure.Cache.debug(s"Node Configuration cache: invalidation request for nodes with action: [${actions.map(_._1).map { _.value }.mkString(",")}]") *>
+        invalidateMergeUpdateSemaphore.withPermit(for {
+          elements     <- invalidateNodeConfigurationRequest.takeAll
+          allActions   =  (elements.flatten ++ actions)
+          _            <- invalidateNodeConfigurationRequest.offer(allActions)
+        } yield ())
+    }
+  }
+  // ? question ?
+  // how to properly ensure that cache is synchro ?
+  // we have the begin date of the nodeexpectedreport that my offer a way to ensure that we don't replace
+  // a value with something older
+
+  // Merge the data with the cache
+  // Strategy:
+  // * if nodeId is not in cache, remove the nodeId
+  // * if the nodeId is in cache, and value is None, augment cache
+  // * if the nodeid is in cache, and value is not None, pick the most recent value to augment cache
+  // Returns the values merged with cache
+  private[this] def mergeDataFromDBWithCache(fromDb: Map[NodeId, Option[NodeExpectedReports]]):IOResult[ Map[NodeId, Option[NodeExpectedReports]] ] = {
+    // In a semaphore, nothing should change the cache
+    semaphore.withPermit(
+      IOResult.effect({
+      val mergedEntries = fromDb.map {  case (nodeId, expFromDb) => cache.get(nodeId) match {
+        case None => None // nodeid not in cache, dropping
```

Comment:
for example, here I don't understand how the nodeId could be not in cache, since we restricted the call to fromDb to only node in cache but not initialized

Reply:
adding/removing nodes can happen at the same time
also, a node may exist without compliance (ignored, or policy not yet generated)

Reply:
For the case without compliance, it's ok: it was none, it's still none after. 
For the case "adding/removing nodes can happen at the same time", I don't understand: here, we are just augmenting the cache knwon entries, we don't remove entries. And there's a semaphore to avoid side effects from `performAction`